python/tchannel/tornado/stream.py

```diff
@@ -0,0 +1,143 @@
+except ImportError:  # pragma: no cover
+    from toro import Condition
+
+
+class ArgStream(object):
+
+    def __init__(self):
+        self._state = StreamState.init
+
+    def read(self):
+        """Async read from internal stream buffer
+
+        if it reaches the end of the stream, it will return empty(""), caller
+        will depend on the return value to tell if it reaches the end of the
+        stream.
+
+        It doesn't support seek functionality, which means it will only read
+        the stream data once in single direction.
+
+        NOTE: The data which has been read will be discarded in the stream.
+
+        :return:chunk of bytes read from stream
+        """
+        raise NotImplementedError()
+
+    def write(self, chunk):
+        """Async write to internal stream buffer
+
+        :return chunk of data in the stream
+
+        :raises StreamingException:
+            if stream has been closed, it will raise StreamingException
+        """
+        raise NotImplementedError()
+
+    def close(self):
+        raise NotImplementedError()
+
+
+class InMemArgStream(ArgStream):
+
+    def __init__(self, buf=None, auto_close=True):
+        """In-Memory based arg stream
+
+        :param buf:
+        :return:
+        """
+        super(InMemArgStream, self).__init__()
+        self._stream = buf or ""
+        self._condition = Condition()
+        self.auto_close = auto_close
+
+    @tornado.gen.coroutine
+    def read(self):
+        if self._state != StreamState.completed and self._stream == "":
+            yield self._condition.wait()
+
+        chunk = self._stream
+        self._stream = ""
+        raise tornado.gen.Return(chunk)
+
+    @tornado.gen.coroutine
+    def write(self, chunk):
+        if self._state == StreamState.completed:
+            raise StreamingException("Stream has been closed.")
+
+        self._stream += chunk
+        self._condition.notify()
+
+    @tornado.gen.coroutine
+    def close(self):
+        self._state = StreamState.completed
+        self._condition.notify()
+
+
+class PipeArgStream(ArgStream):
+
+    def __init__(self, rpipe, wpipe=None, auto_close=False):
+        """Pipe-based arg stream
+
+        NOTE: reading from or writing to files, use os.open to get the file
+        descriptor instead of python's open. Socket file descriptors and
+        others are fine.
+
+        when you use os.pipe to generate one write pipe and one read pipe, you
+        need to pass both of them into init method.
+
+        :param rpipe: an integer file descriptor which supports read ops
+        :param wpipe: an integer file descriptor which supports write ops
+        :param auto: flag to indicate to close the stream automatically or not
+        """
+        super(PipeArgStream, self).__init__()
+        self._rpipe = rpipe
+        self._wpipe = wpipe
+
+        self._rs = (PipeIOStream(self._rpipe) if
+                    self._rpipe is not None else None)
```

Comment:
Since `rpipe` is required, don't allow it to be None. Assert that it is not None.